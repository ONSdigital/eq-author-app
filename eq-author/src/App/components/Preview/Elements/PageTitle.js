import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import Error from "App/components/Preview/Error/Error";

const Title = styled.h1`
  font-size: 1.4em;
  margin: 0 0 1em;
`;

const PageTitle = ({ title }) => {
  let pageTitle = title && title.replace(/(<p[^>]+?>|<p>|<\/p>)/gim, "");
  return (
    <Title>
      {pageTitle ? (
        /*  eslint-disable-next-line react/no-danger */
        <div dangerouslySetInnerHTML={{ __html: pageTitle }} />
      ) : (
        <Error data-test="no-title" large>
          Missing Page Title
        </Error>
      )}
    </Title>
  );
};

PageTitle.propTypes = {
  title: PropTypes.string
};

export default PageTitle;
