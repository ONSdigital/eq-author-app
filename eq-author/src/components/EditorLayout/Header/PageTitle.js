import React from "react";
import PropTypes from "prop-types";

import styled from "styled-components";
import { colors } from "constants/theme";

const Wrapper = styled.div`
  display: flex;
  align-items: center;
  color: ${colors.white};
  padding: 0 1.5em 1em;
`;

const Title = styled.h1`
  font-size: 1.5em;
  font-weight: 600;
  margin: 0;
  width: 100%;
  line-height: 1.3;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`;

const PageTitle = ({ children }) => (
  <Wrapper>
    <Title data-test="questionnaire-title">{children}&nbsp;</Title>
  </Wrapper>
);

PageTitle.propTypes = {
  children: PropTypes.node,
};

export default PageTitle;
