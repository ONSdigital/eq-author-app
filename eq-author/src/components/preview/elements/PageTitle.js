import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import Error from "components/preview/Error";

const Title = styled.h1`
  font-size: 1.4em;
  margin: 0 0 1em;
  word-wrap: break-word;
`;

const PageTitle = ({ title, missingText = "Missing Page Title" }) => {
  let pageTitle = title && title.replace(/(<p[^>]+?>|<p>|<\/p>)/gim, "");
  return (
    <Title data-test="page-title">
      {pageTitle ? (
        /*  eslint-disable-next-line react/no-danger */
        <div dangerouslySetInnerHTML={{ __html: pageTitle }} />
      ) : (
        <Error data-test="no-title" large>
          {missingText}
        </Error>
      )}
    </Title>
  );
};

PageTitle.propTypes = {
  title: PropTypes.string,
  missingText: PropTypes.string,
};

export default PageTitle;
