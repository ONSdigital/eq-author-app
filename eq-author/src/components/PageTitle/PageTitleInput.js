import React, { useState } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { Field, Label } from "components/Forms";
import { colors, radius } from "constants/theme";
import ValidationError from "components/ValidationError";
// import ExternalLink from "components-themed/ExternalLink";

import PasteModal from "components/modals/PasteModal";
import { reduceMultipleSpaces } from "utils/reduceMultipleSpaces";

const PageTitleContent = styled.div`
  color: ${colors.black};
`;

const Heading = styled.h2`
  font-size: 1em;
  font-weight: bold;
  line-height: 1.3em;
  margin-bottom: 0.4em;
`;

const PageDescriptionContent = styled.div`
  margin-bottom: 1.3em;
`;

const StyledInput = styled.input`
  width: 100%;
  font-size: 1em;
  color: ${colors.black};
  border: thin solid ${colors.grey};
  border-radius: ${radius};
  padding: 0.5em;
  outline: thin solid transparent;
  ${({ hasError }) =>
    hasError &&
    `
    border-color: ${colors.errorPrimary};
    &:focus,
    &:focus-within {
      border-color: ${colors.errorPrimary};
      outline-color: ${colors.errorPrimary};
      box-shadow: 0 0 0 2px ${colors.errorPrimary};
    }
    &:hover {
      border-color: ${colors.errorPrimary};
      outline-color: ${colors.errorPrimary};
    }
  `}
`;

const PageTitleInput = ({
  heading,
  onChange,
  onUpdate,
  pageDescription,
  inputTitlePrefix,
  altFieldName,
  errorMessage,
}) => {
  const [pasteModalInfo, setPasteModalInfo] = useState({
    show: false,
    text: "",
    event: {},
  });

  const handleChange = (e) => {
    e.target.value = e.target.value.replace(/\n/g, " ");
    e.target.value = e.target.value.replace(/\s+/g, " ");
    onChange(e.target);
  };

  const handlePaste = (event) => {
    const text = event.clipboardData.getData("text");
    event.persist();
    if (/\s{2,}/g.test(text)) {
      event.preventDefault();
      setPasteModalInfo({
        show: true,
        text: text,
        event: event,
      });
    }
  };

  const updateOnPaste = () => {
    const { text, event } = pasteModalInfo;
    if (event && event.persist) {
      const target = event.target;
      const cursorPosition = target.selectionStart;
      const currentValue = target.value;

      // Insert the pasted text at the cursor position
      const newValue =
        currentValue.substring(0, cursorPosition) +
        reduceMultipleSpaces(text) +
        currentValue.substring(target.selectionEnd);

      const updatedEvent = { ...event, persist: undefined }; // Create a new event without the persist method
      updatedEvent.target.value = newValue.trim().replace(/\s+/g, " ");

      onChange(updatedEvent.target);
    }

    // Clear the pasteModalInfo state
    setPasteModalInfo({ show: false, text: "" });
  };

  const cancelPaste = () => {
    setPasteModalInfo({ show: false, text: "" });
  };
  return (
    <PageTitleContent>
      <PasteModal
        isOpen={pasteModalInfo.show}
        onConfirm={updateOnPaste}
        onCancel={cancelPaste}
      />
      <Heading data-test="page-title-input-heading">{heading}</Heading>
      <PageDescriptionContent>
        The page title is the first thing read by screen readers and helps users
        of assistive technology understand what the page is about. It is shown
        in the browser&apos;s title bar or in the page&apos;s tab. Page titles
        follow the structure: &apos;page description - questionnaire
        title&apos;.
      </PageDescriptionContent>
      {/* From interaction designer - this will be brought back when the page titles link is fixed */}
      {/* <p>
        For help writing a page description, see our{" "}
        <ExternalLink
          url="https://ons-design-system.netlify.app/guidance/page-titles-and-urls/#page-titles"
          linkText="design system guidance on page titles"
        />
      </p> */}
      <Field data-test="page-title-missing-error">
        <Label
          data-test="page-title-input-field-label"
          htmlFor={altFieldName ? altFieldName : "pageDescription"}
        >
          {inputTitlePrefix
            ? `${inputTitlePrefix} page description`
            : "Page description"}
        </Label>
        <StyledInput
          id={altFieldName ? altFieldName : "pageDescription"}
          type="text"
          data-test="txt-page-description"
          autoComplete="off"
          name={altFieldName ? altFieldName : "pageDescription"}
          placeholder=""
          onPaste={(e) => handlePaste(e)}
          onChange={(e) => handleChange(e)}
          onBlur={(e) => onUpdate(e.target)}
          value={pageDescription || ""}
          hasError={errorMessage}
        />
        {errorMessage && <ValidationError>{errorMessage}</ValidationError>}
      </Field>
    </PageTitleContent>
  );
};

PageTitleInput.propTypes = {
  pageDescription: PropTypes.string,
  inputTitlePrefix: PropTypes.string,
  heading: PropTypes.string,
  altFieldName: PropTypes.string,
  errorMessage: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  onUpdate: PropTypes.func.isRequired,
};

PageTitleInput.defaultProps = {
  heading: "Page title and description",
};

export default PageTitleInput;
