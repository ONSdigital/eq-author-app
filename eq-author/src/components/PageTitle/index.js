import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { find } from "lodash";

import PageTitleInput from "./PageTitleInput";

import { pageDescriptionErrors } from "constants/validationMessages";

const PageTitleSegment = styled.div`
  margin-top: 2em;
`;

const PageTitleContainer = ({
  heading,
  marginless,
  pageDescription,
  inputTitlePrefix,
  altFieldName,
  altError,
  errors,
  onChange,
  onUpdate,
}) => {
  const pageDescriptionError =
    altFieldName && altError
      ? find(errors, { errorCode: altError }) ||
        find(errors, {
          errorCode: "ERR_UNIQUE_PAGE_DESCRIPTION",
        })
      : find(errors, { errorCode: "PAGE_DESCRIPTION_MISSING" }) ||
        find(errors, {
          errorCode: "ERR_UNIQUE_PAGE_DESCRIPTION",
        });

  const allErrorCodes = Object.keys(pageDescriptionErrors);
  const errorCode = allErrorCodes.find(
    (code) => pageDescriptionError?.errorCode === code
  );

  const errorMessage = pageDescriptionErrors[errorCode];
  return (
    <div
      title="Page description"
      className="pageDescription"
      dataTestIdPrefix="page-title-"
      defaultOpen
      withoutHideThis
      variant={marginless ? "marginlessContent" : "content"}
    >
      <PageTitleSegment data-test="page-title-container">
        <PageTitleInput
          heading={heading}
          pageDescription={pageDescription}
          inputTitlePrefix={inputTitlePrefix}
          altFieldName={altFieldName}
          onUpdate={onUpdate}
          onChange={onChange}
          errorMessage={errorMessage}
        />
      </PageTitleSegment>
    </div>
  );
};

PageTitleContainer.propTypes = {
  heading: PropTypes.string,
  marginless: PropTypes.bool,
  pageDescription: PropTypes.string,
  inputTitlePrefix: PropTypes.string,
  altFieldName: PropTypes.string,
  altError: PropTypes.string,
  errors: PropTypes.arrayOf(PropTypes.object),
  onChange: PropTypes.func.isRequired,
  onUpdate: PropTypes.func.isRequired,
};

export default PageTitleContainer;
