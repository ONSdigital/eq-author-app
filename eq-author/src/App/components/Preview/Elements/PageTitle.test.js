import { shallow } from "enzyme";
import React from "react";

import PageTitle from "./PageTitle";

describe("PageTitle", () => {
  it("should render", () => {
    const wrapper = shallow(<PageTitle title="foobar" />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should render a warning there is no title", () => {
    const wrapper = shallow(<PageTitle title="" />);
    expect(wrapper.find('[data-test="no-title"]')).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });
});
