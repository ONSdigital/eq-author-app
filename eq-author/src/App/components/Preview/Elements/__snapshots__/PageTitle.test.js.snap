// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageTitle should render 1`] = `
<PageTitle__Title>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "foobar",
      }
    }
  />
</PageTitle__Title>
`;

exports[`PageTitle should render a warning there is no title 1`] = `
<PageTitle__Title>
  <Error
    data-test="no-title"
    large={true}
    margin={true}
  >
    Missing Page Title
  </Error>
</PageTitle__Title>
`;
